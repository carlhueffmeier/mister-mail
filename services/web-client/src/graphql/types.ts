/* tslint:disable */
/* eslint-disable */
//  This file was automatically generated and should not be edited.

export type CreateCampaignInput = {
  name: string,
  questionText: string,
  destinations: Array< DestinationInput >,
};

export type DestinationInput = {
  name: string,
  email: string,
};

export type NotifyCampaignUpdateInput = {
  id?: string | null,
  uid?: string | null,
  created?: string | null,
  updated?: string | null,
  name?: string | null,
  from?: string | null,
  questionText?: string | null,
  destinations?: Array< DestinationInput | null > | null,
  stats?: CampaignStatsInput | null,
};

export type CampaignStatsInput = {
  Created?: number | null,
  Sent?: number | null,
  Delivered?: number | null,
  Complaint?: number | null,
  Rejected?: number | null,
  Bounce?: number | null,
  Opened?: number | null,
  Responded?: number | null,
};

export type CreateCampaignMutationVariables = {
  data: CreateCampaignInput,
};

export type CreateCampaignMutation = {
  createCampaign:  {
    __typename: "Campaign",
    id: string,
    uid: string,
    created: string,
    updated: string,
    name: string,
    from: string,
    questionText: string,
    destinations:  Array< {
      __typename: "Destination",
      name: string,
      email: string,
    } >,
    stats:  {
      __typename: "CampaignStats",
      Created: number | null,
      Sent: number | null,
      Delivered: number | null,
      Complaint: number | null,
      Rejected: number | null,
      Bounce: number | null,
      Opened: number | null,
      Responded: number | null,
    },
  },
};

export type NotifyCampaignUpdateMutationVariables = {
  data: NotifyCampaignUpdateInput,
};

export type NotifyCampaignUpdateMutation = {
  notifyCampaignUpdate:  {
    __typename: "Campaign",
    id: string,
    uid: string,
    created: string,
    updated: string,
    name: string,
    from: string,
    questionText: string,
    destinations:  Array< {
      __typename: "Destination",
      name: string,
      email: string,
    } >,
    stats:  {
      __typename: "CampaignStats",
      Created: number | null,
      Sent: number | null,
      Delivered: number | null,
      Complaint: number | null,
      Rejected: number | null,
      Bounce: number | null,
      Opened: number | null,
      Responded: number | null,
    },
  } | null,
};

export type GetCampaignsQuery = {
  getCampaigns:  Array< {
    __typename: "Campaign",
    id: string,
    uid: string,
    created: string,
    updated: string,
    name: string,
    from: string,
    questionText: string,
    destinations:  Array< {
      __typename: "Destination",
      name: string,
      email: string,
    } >,
    stats:  {
      __typename: "CampaignStats",
      Created: number | null,
      Sent: number | null,
      Delivered: number | null,
      Complaint: number | null,
      Rejected: number | null,
      Bounce: number | null,
      Opened: number | null,
      Responded: number | null,
    },
  } >,
};

export type CampaignUpdateSubscription = {
  campaignUpdate:  {
    __typename: "Campaign",
    id: string,
    uid: string,
    created: string,
    updated: string,
    name: string,
    from: string,
    questionText: string,
    destinations:  Array< {
      __typename: "Destination",
      name: string,
      email: string,
    } >,
    stats:  {
      __typename: "CampaignStats",
      Created: number | null,
      Sent: number | null,
      Delivered: number | null,
      Complaint: number | null,
      Rejected: number | null,
      Bounce: number | null,
      Opened: number | null,
      Responded: number | null,
    },
  } | null,
};
