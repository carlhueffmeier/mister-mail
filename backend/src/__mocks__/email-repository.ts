import { DynamoDB } from 'aws-sdk';
import Log from '@dazn/lambda-powertools-logger';
import { EmailDynamoDbRecord } from '../types';
import { EmailStatus } from '../types';
import { Email } from '../types';
import { getEpochSeconds } from '../../lib/utils';

export * from '../types';
export class EmailRepository {
  constructor(_options: {
    dynamoDbDocumentClient: DynamoDB.DocumentClient;
    tableName: string;
    logger: Log;
  }) {
    // Noop
  }

  async create(createData: Readonly<Email>): Promise<EmailDynamoDbRecord> {
    const timestamp = getEpochSeconds();
    return {
      pk: createData.messageId,
      sk: `mail`,
      uid: createData.uid,
      campaignId: createData.campaignId,
      messageId: createData.messageId,
      created: timestamp,
      updated: timestamp,
      status: createData.status,
    };
  }

  async updateStatus(_updateParams: {
    uid: string;
    campaignId: string;
    messageId: string;
    status: EmailStatus;
  }): Promise<void> {
    // Noop
  }

  async findOne(_findParams: {
    uid: string;
    campaignId: string;
    messageId: string;
  }): Promise<EmailDynamoDbRecord> {
    return {} as EmailDynamoDbRecord;
  }
}
