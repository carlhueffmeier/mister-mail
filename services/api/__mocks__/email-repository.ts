import { DynamoDB } from 'aws-sdk';
import Log from '@dazn/lambda-powertools-logger';
import {
  EmailStatus,
  Email,
  EmailDynamoDbRecord,
} from '../email-repository.types';

export class EmailRepository {
  constructor(_options: {
    dynamoDbDocumentClient: DynamoDB.DocumentClient;
    tableName: string;
    logger: Log;
  }) {
    // Noop
  }

  async create(createData: Readonly<Email>): Promise<EmailDynamoDbRecord> {
    const timestamp = Date.now();
    return {
      pk: createData.messageId,
      sk: `mail`,
      messageId: createData.messageId,
      created: timestamp,
      updated: timestamp,
      status: createData.status,
    };
  }

  async updateStatus(_messageId: string, _status: EmailStatus): Promise<void> {
    // TODO
  }
}
