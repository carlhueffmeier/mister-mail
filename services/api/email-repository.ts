import { DynamoDB } from 'aws-sdk';
import Log from '@dazn/lambda-powertools-logger';
import {
  EmailStatus,
  Email,
  EmailDynamoDbRecord,
} from './email-repository.types';

export * from './email-repository.types';
export class EmailRepository {
  dynamoDbDocumentClient: DynamoDB.DocumentClient;
  tableName: string;
  logger: Log;

  constructor(options: {
    dynamoDbDocumentClient: DynamoDB.DocumentClient;
    tableName: string;
    logger: Log;
  }) {
    this.dynamoDbDocumentClient = options.dynamoDbDocumentClient;
    this.tableName = options.tableName;
    this.logger = options.logger;
  }

  async create(createData: Readonly<Email>): Promise<EmailDynamoDbRecord> {
    const timestamp = Date.now();
    const newItem: EmailDynamoDbRecord = {
      pk: createData.messageId,
      sk: `mail`,
      messageId: createData.messageId,
      created: timestamp,
      updated: timestamp,
      status: createData.status,
    };
    const putParams: DynamoDB.DocumentClient.PutItemInput = {
      TableName: this.tableName,
      Item: newItem,
    };
    this.logger.debug('Creating new email record', { createData, putParams });
    await this.dynamoDbDocumentClient.put(putParams).promise();
    return newItem;
  }

  async updateStatus(_messageId: string, _status: EmailStatus): Promise<void> {
    // TODO
  }
}
