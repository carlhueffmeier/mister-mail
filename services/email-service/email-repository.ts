import { DynamoDB } from 'aws-sdk';
import Log from '@dazn/lambda-powertools-logger';
import {
  EmailStatus,
  Email,
  EmailDynamoDbRecord,
} from './email-repository.types';

export * from './email-repository.types';
export class EmailRepository {
  dynamoDbDocumentClient: DynamoDB.DocumentClient;
  tableName: string;
  logger: Log;

  constructor(options: {
    dynamoDbDocumentClient: DynamoDB.DocumentClient;
    tableName: string;
    logger: Log;
  }) {
    this.dynamoDbDocumentClient = options.dynamoDbDocumentClient;
    this.tableName = options.tableName;
    this.logger = options.logger;
  }

  async create(createData: Readonly<Email>): Promise<EmailDynamoDbRecord> {
    const timestamp = Date.now();
    const newItem: EmailDynamoDbRecord = {
      pk: createData.messageId,
      sk: `mail`,
      messageId: createData.messageId,
      created: timestamp,
      updated: timestamp,
      status: createData.status,
    };
    const putParams: DynamoDB.DocumentClient.PutItemInput = {
      TableName: this.tableName,
      Item: newItem,
    };
    this.logger.debug('Creating new email record', { createData, putParams });
    await this.dynamoDbDocumentClient.put(putParams).promise();
    return newItem;
  }

  async updateStatus(messageId: string, status: EmailStatus): Promise<void> {
    const updateParams = {
      TableName: this.tableName,
      Key: { pk: messageId, sk: 'mail' },
      UpdateExpression: 'set #status = :new_status',
      ExpressionAttributeNames: {
        '#status': 'status',
      },
      ExpressionAttributeValues: {
        ':new_status': status,
      },
    };
    this.logger.debug('Updating email record', { updateParams });
    await this.dynamoDbDocumentClient.update(updateParams).promise();
  }
}
