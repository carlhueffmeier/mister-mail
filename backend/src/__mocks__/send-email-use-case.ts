import AWS from 'aws-sdk';
import { EmailRepository } from '../email-repository';
import { Email, EmailStatus } from '../types';
import { SendEmailRequest } from '../types';
import Log from '@dazn/lambda-powertools-logger';

export class SendEmailUseCase {
  constructor(_args: {
    _ses: AWS.SES;
    _emailRepository: EmailRepository;
    _configurationSet: string;
    _logger: Log;
  }) {
    // Noop
  }

  async sendEmail(_request: SendEmailRequest): Promise<Email> {
    return {
      uid: 'uid',
      campaignId: 'campaignId',
      messageId: 'messageId',
      status: EmailStatus.Sent,
    };
  }

  async initializeEventDestinations(): Promise<void> {
    // Noop
  }
}
