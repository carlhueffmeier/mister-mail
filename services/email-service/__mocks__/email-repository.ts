import { DynamoDB } from 'aws-sdk';
import Log from '@dazn/lambda-powertools-logger';
import { EmailDynamoDbRecord } from '../email-repository.types';
import { EmailStatus } from '../../../lib/types';
import { Email } from '../../../lib/types';

export * from '../email-repository.types';
export class EmailRepository {
  constructor(_options: {
    dynamoDbDocumentClient: DynamoDB.DocumentClient;
    tableName: string;
    logger: Log;
  }) {
    // Noop
  }

  async create(createData: Readonly<Email>): Promise<EmailDynamoDbRecord> {
    const timestamp = Date.now();
    return {
      pk: createData.messageId,
      sk: `mail`,
      uid: createData.uid,
      campaignId: createData.campaignId,
      messageId: createData.messageId,
      created: timestamp,
      updated: timestamp,
      status: createData.status,
    };
  }

  async updateStatus(_messageId: string, _status: EmailStatus): Promise<void> {
    // Noop
  }

  async findById(_messageId: string): Promise<EmailDynamoDbRecord> {
    return {} as EmailDynamoDbRecord;
  }
}
