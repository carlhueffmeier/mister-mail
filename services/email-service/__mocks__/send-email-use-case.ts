import AWS from 'aws-sdk';
import { EmailRepository } from '../email-repository';
import { Email, EmailStatus } from '../../../lib/types';
import { SendEmailRequest } from '../send-email-use-case.types';
import Log from '@dazn/lambda-powertools-logger';

export class SendEmailUseCase {
  constructor(_args: {
    _ses: AWS.SES;
    _emailRepository: EmailRepository;
    _configurationSet: string;
    _logger: Log;
  }) {
    // Noop
  }

  async sendEmail(_request: SendEmailRequest): Promise<Email> {
    return {
      uid: 'uid',
      campaignId: 'campaignId',
      messageId: 'messageId',
      status: EmailStatus.Sent,
    };
  }

  async initializeEventDestinations(): Promise<void> {
    // Noop
  }
}
