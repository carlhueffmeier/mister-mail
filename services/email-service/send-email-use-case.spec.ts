import AWS from 'aws-sdk';
import { SendEmailUseCase } from './send-email-use-case';
import { SendEmailRequest } from './send-email-use-case.types';
import { EmailRepository } from './email-repository';
import Log from '@dazn/lambda-powertools-logger';

jest.mock('../../lib/utils');

describe('send-email-use-case', () => {
  const fakeEmailRepository = ({
    create: jest.fn(),
  } as unknown) as EmailRepository;
  const fakeSes = ({ sendEmail: jest.fn() } as unknown) as AWS.SES;
  const configurationSet = 'configurationSet';

  describe('sendEmail', () => {
    describe('given valid request', () => {
      const request: SendEmailRequest = {
        subject: 'subject',
        text: 'text',
        html: 'html',
        source: 'source',
        destination: 'destination',
      };
      it('should send the email', async () => {
        const useCase = new SendEmailUseCase({
          ses: fakeSes,
          emailRepository: fakeEmailRepository,
          configurationSet,
          logger: ({ debug: jest.fn() } as unknown) as Log,
        });
        const fakeResponse = {
          async promise(): Promise<AWS.SES.SendEmailResponse> {
            return { MessageId: 'abc' } as AWS.SES.SendEmailResponse;
          },
        };
        (fakeSes.sendEmail as jest.Mock).mockReturnValueOnce(fakeResponse);

        await useCase.sendEmail(request);

        expect((fakeSes.sendEmail as jest.Mock).mock.calls).toMatchSnapshot();
      });

      it('should save the email', async () => {
        const useCase = new SendEmailUseCase({
          ses: fakeSes,
          emailRepository: fakeEmailRepository,
          configurationSet,
          logger: ({ debug: jest.fn() } as unknown) as Log,
        });
        const fakeResponse = {
          async promise(): Promise<AWS.SES.SendEmailResponse> {
            return { MessageId: 'abc' } as AWS.SES.SendEmailResponse;
          },
        };
        (fakeSes.sendEmail as jest.Mock).mockReturnValueOnce(fakeResponse);

        await useCase.sendEmail(request);

        expect(
          (fakeEmailRepository.create as jest.Mock).mock.calls,
        ).toMatchSnapshot();
      });
    });
  });
});
