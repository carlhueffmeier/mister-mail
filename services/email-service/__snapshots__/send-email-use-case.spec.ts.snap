// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`send-email-use-case sendEmail given valid request should save the email 1`] = `
Array [
  Array [
    Object {
      "campaignId": "campaignId",
      "messageId": "abc",
      "status": "Sent",
      "uid": "uid",
    },
  ],
  Array [
    Object {
      "campaignId": "campaignId",
      "messageId": "abc",
      "status": "Sent",
      "uid": "uid",
    },
  ],
]
`;

exports[`send-email-use-case sendEmail given valid request should send the email 1`] = `
Array [
  Array [
    Object {
      "ConfigurationSetName": "configurationSet",
      "Destination": Object {
        "ToAddresses": Array [
          "destination",
        ],
      },
      "Message": Object {
        "Body": Object {
          "Html": Object {
            "Data": "html",
          },
          "Text": Object {
            "Data": "text",
          },
        },
        "Subject": Object {
          "Data": "subject",
        },
      },
      "Source": "source",
    },
  ],
]
`;
